File: js/engine/EmotionalAnalyzer.js
/**
 * EmotionalAnalyzer.js - Agente de Procesamiento Emocional y Acústico
 * Analiza prosodia, ritmo y vocabulario afectivo para detectar ansiedad, tristeza o agitación.
 */

export class EmotionalAnalyzer {
  static analyzeAffectivity(inputText, speechRate = 1.0, pausesCount = 0) {
    const textLower = inputText.toLowerCase();

    let valence = 0.8; // 1.0 = Alegre, 0.5 = Neutro, 0.0 = Agitado/Ansioso
    let detectedState = 'Estable / Tranquilo';
    let isAgitatedOrDisoriented = false;

    // Palabras clave de ansiedad / tristeza / desorientación
    const anxietyKeywords = ['miedo', 'asustada', 'angustia', 'sola', 'triste', 'perdida', 'dónde', 'donde', 'ayuda'];
    const positiveKeywords = ['lindo', 'alegría', 'gracias', 'hermoso', 'bien', 'contenta', 'te quiero'];

    let matchesAnxiety = 0;
    let matchesPositive = 0;

    anxietyKeywords.forEach(word => {
      if (textLower.includes(word)) matchesAnxiety++;
    });

    positiveKeywords.forEach(word => {
      if (textLower.includes(word)) matchesPositive++;
    });

    if (matchesAnxiety > 0 || speechRate > 1.4 || pausesCount > 3) {
      valence = Math.max(0.2, 0.7 - matchesAnxiety * 0.2);
      detectedState = matchesAnxiety > 1 ? 'Agitación / Desorientación' : 'Leve Ansiedad / Tristeza';
      isAgitatedOrDisoriented = true;
    } else if (matchesPositive > 0) {
      valence = Math.min(1.0, 0.85 + matchesPositive * 0.1);
      detectedState = 'Alegre / Receptivo';
    }

    return {
      valence,
      detectedState,
      isAgitatedOrDisoriented,
      acousticMetrics: {
        speechRate,
        pausesCount,
        intensity: isAgitatedOrDisoriented ? 'Elevada' : 'Normal'
      }
    };
  }
}

File: backend/agents/emotional_analyzer.py
"""emotional_analyzer.py - Port de EmotionalAnalyzer.js.

Heurística de valencia/estado afectivo a partir del texto reconocido (y, opcionalmente,
métricas de habla que en esta versión web-based no llegan del navegador todavía).
"""
from dataclasses import dataclass, field

ANXIETY_KEYWORDS = ["miedo", "asustada", "angustia", "sola", "triste", "perdida", "dónde", "donde", "ayuda"]
POSITIVE_KEYWORDS = ["lindo", "alegría", "gracias", "hermoso", "bien", "contenta", "te quiero"]


@dataclass
class EmotionalAnalysis:
    valence: float
    detected_state: str
    is_agitated_or_disoriented: bool
    speech_rate: float = 1.0
    pauses_count: int = 0
    intensity: str = "Normal"


def analyze_affectivity(input_text: str, speech_rate: float = 1.0, pauses_count: int = 0) -> EmotionalAnalysis:
    text_lower = input_text.lower()

    valence = 0.8
    detected_state = "Estable / Tranquilo"
    is_agitated_or_disoriented = False

    matches_anxiety = sum(1 for word in ANXIETY_KEYWORDS if word in text_lower)
    matches_positive = sum(1 for word in POSITIVE_KEYWORDS if word in text_lower)

    if matches_anxiety > 0 or speech_rate > 1.4 or pauses_count > 3:
        valence = max(0.2, 0.7 - matches_anxiety * 0.2)
        detected_state = "Agitación / Desorientación" if matches_anxiety > 1 else "Leve Ansiedad / Tristeza"
        is_agitated_or_disoriented = True
    elif matches_positive > 0:
        valence = min(1.0, 0.85 + matches_positive * 0.1)
        detected_state = "Alegre / Receptivo"

    return EmotionalAnalysis(
        valence=valence,
        detected_state=detected_state,
        is_agitated_or_disoriented=is_agitated_or_disoriented,
        speech_rate=speech_rate,
        pauses_count=pauses_count,
        intensity="Elevada" if is_agitated_or_disoriented else "Normal",
    )
